api-server/lib/AirSwap.js

```
@@ -456,6 +474,77 @@ class AirSwap {
       { gasLimit, gasPrice },
     )
   }
+
+  // registers a new PGP keyset on the contract for this wallet
+  async registerPGPKey(){
+    const address = this.wallet.address
+    this.signedSeed = this.wallet.signMessage(`I'm generating my encryption keys for AirSwap ${address}`)
+
+    if(this.pgpKeys[address]) {
+      return new Error('Key already set on PGP contract, should only be set once')
+    }
+
+    const {privateKeyArmored, publicKeyArmored} = await openpgp.generateKey({
+      userIds: [{ address }],
+      curve: 'p256', // ECC curve name, most widely supported
+      passphrase: this.signedSeed,
+    })
+    const walletPGPKey = {privateKeyArmored, publicKeyArmored}
+    this.pgpKeys[address] = walletPGPKey
+    const ipfsHash = await ipfs.add(JSON.stringify(walletPGPKey))
+    return this.pgpContract.addPublicKey(ipfsHash, {
+      value: 0,
+      gasLimit: 160000,
+      gasPrice: utils.parseEther('0.000000040'),
+    })
+  }
+
+  // look up a PGP keyset by wallet address
+  async getPGPKey(address){
+    if(this.pgpKeys[address]){
+      return this.pgpKeys[address]
+    } else {
+      const ipfsHash = await this.pgpContract.addressToPublicKey(this.wallet.address)
+      this.pgpKeys[address] = await ipfs.cat(ipfsHash)
+      return this.pgpKeys[address]
```

this can fail if wallet address is not registered right? should we return an error in that case?

---

Yes, changed. In terms of communicating that to whatever is calling the API server, I'm not sure how we should be handling that in general. Probably should use error codes (like responses with status `500` when an error occurs, instead of the `200` we currently always send). But that's outside of the scope of this PR.